File: CuiYanling/Longest Substring Without Repeating Characters.java
public class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character,Integer> pos = new HashMap<Character,Integer>();
        int start = -1,ret = 0;
        for(int i=0;i<s.length();i++){
            Character c = s.charAt(i);
            if(pos.containsKey(c))
                start = Math.max(start,pos.get(c));
            
            pos.put(c,i);
            ret = Math.max(ret,i-start);
        }
        return ret;
    }
}

File: TangHongyin/longest_substring_without_repeating_characters.py
class Solution(object):

    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        dict = {}
        start = 0
        length = 0
        for i in range(len(s)):
            if s[i] in dict and dict[s[i]] != -1:
                length = max(length, i - start)
                for j in range(start, dict[s[i]]):
                    dict[s[j]] = -1
                start = dict[s[i]] + 1
            dict[s[i]] = i
        if start < len(s):
            length = max(length, len(s) - start)
        return length


if __name__ == "__main__":
    str = "abba"
    solution = Solution()
    print solution.lengthOfLongestSubstring(str)
